### PR TITLE
[WIP][9.0] business_requirement

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
This PR contains the business requirement module for v9.0.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/OCA/business-requirement/pull/10%23issuecomment-225416384%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717584%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717587%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717592%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717606%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717616%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717660%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717785%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23issuecomment-227422988%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23issuecomment-228965635%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/10%23issuecomment-247374791%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23issuecomment-225416384%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40jbeficent%20Thanks%20for%20this%20PR.%20I%20think%20we%20might%20want%20to%20finish%20all%20rework%20in%20v8%20before%20we%20merge%20this%20one%20because%20many%20topics%20might%20apply%20in%20both%20cases%22%2C%20%22created_at%22%3A%20%222016-06-12T08%3A15%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20agree.%20This%20PR%20should%20be%20rebased%20on%20the%208.0%20final%20version%2C%20once%20its%20merged.%22%2C%20%22created_at%22%3A%20%222016-06-21T12%3A21%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1246629%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dreispt%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40jbeficent%20could%20you%20review%20the%20work%20done%20in%20current%20branch%208.0%20for%20business_requirement%20merged%20module%20and%20adapt%20this%20PR%20accordingly%3F%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222016-06-28T06%3A51%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40jbeficent%20I%20close%20this%20PR%20as%20the%20structure%20has%20greatly%20evolved%20in%20v8%20since%20this%20initial%20PR.%5Cr%5CnWe%20should%20start%20a%20migration%20process%20at%20this%20point%20of%20the%20v8%20repo.%22%2C%20%22created_at%22%3A%20%222016-09-15T16%3A12%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20005bfa7ac3b6ec97e0db347241dfbd348300f94f%20business_requirement/models/business.py%2020%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717660%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22normally%20this%20case%20is%20impossible%20%28we%20have%20have%20the%20discussion%20in%20v8%29%22%2C%20%22created_at%22%3A%20%222016-06-12T07%3A56%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement/models/business.py%3AL1-437%22%7D%2C%20%22Pull%20005bfa7ac3b6ec97e0db347241dfbd348300f94f%20business_requirement/README.rst%2012%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717584%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20have%20pending%20for%20v8%20a%20remodeling%20of%20the%20intro%20of%20the%20README.%20Later%20on%20we%20might%20adapt%20it%20to%20v9%20as%20well%22%2C%20%22created_at%22%3A%20%222016-06-12T07%3A45%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement/README.rst%3AL1-286%22%7D%2C%20%22Pull%20005bfa7ac3b6ec97e0db347241dfbd348300f94f%20business_requirement/wizards/mail_compose_message.py%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717785%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22same%20issue%20here%3A%20we%20should%20not%20rewrite%20this%20function%22%2C%20%22created_at%22%3A%20%222016-06-12T08%3A14%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement/wizards/mail_compose_message.py%3AL1-88%22%7D%2C%20%22Pull%20005bfa7ac3b6ec97e0db347241dfbd348300f94f%20business_requirement/README.rst%20271%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717592%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22add%20your%20name%22%2C%20%22created_at%22%3A%20%222016-06-12T07%3A46%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement/README.rst%3AL1-286%22%7D%2C%20%22Pull%20005bfa7ac3b6ec97e0db347241dfbd348300f94f%20business_requirement/README.rst%20250%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717587%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22simplify%20following%20latest%20README%20template%22%2C%20%22created_at%22%3A%20%222016-06-12T07%3A46%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement/README.rst%3AL1-286%22%7D%2C%20%22Pull%20005bfa7ac3b6ec97e0db347241dfbd348300f94f%20business_requirement/models/business.py%20226%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717616%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22in%20v8%20this%20is%20pending%20rewriting/refactoring%20%28currently%20not%20good%20to%20rewrite%20the%20whole%20function%20in%20this%20module%29%22%2C%20%22created_at%22%3A%20%222016-06-12T07%3A49%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement/models/business.py%3AL1-437%22%7D%2C%20%22Pull%20005bfa7ac3b6ec97e0db347241dfbd348300f94f%20business_requirement/models/business.py%20179%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/10%23discussion_r66717606%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20wonder%20whether%20we%20should%20be%20more%20generic%20as%20%5C%22customer%20approval%5C%22%20is%20for%20outside%20projects.%5Cr%5Cnsomething%20like%20%5C%22Internal%20Approval%5C%22%20and%20%5C%22Stakeholder%20Approval%5C%22%22%2C%20%22created_at%22%3A%20%222016-06-12T07%3A49%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement/models/business.py%3AL1-437%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 005bfa7ac3b6ec97e0db347241dfbd348300f94f business_requirement/README.rst 12'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#discussion_r66717584'>File: business_requirement/README.rst:L1-286</a></b>
- <a href='https://github.com/elicoidal'></a> I have pending for v8 a remodeling of the intro of the README. Later on we might adapt it to v9 as well
- [ ] <a href='#crh-comment-Pull 005bfa7ac3b6ec97e0db347241dfbd348300f94f business_requirement/README.rst 250'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#discussion_r66717587'>File: business_requirement/README.rst:L1-286</a></b>
- <a href='https://github.com/elicoidal'></a> simplify following latest README template
- [ ] <a href='#crh-comment-Pull 005bfa7ac3b6ec97e0db347241dfbd348300f94f business_requirement/README.rst 271'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#discussion_r66717592'>File: business_requirement/README.rst:L1-286</a></b>
- <a href='https://github.com/elicoidal'></a> add your name
- [ ] <a href='#crh-comment-Pull 005bfa7ac3b6ec97e0db347241dfbd348300f94f business_requirement/models/business.py 179'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#discussion_r66717606'>File: business_requirement/models/business.py:L1-437</a></b>
- <a href='https://github.com/elicoidal'></a> I wonder whether we should be more generic as "customer approval" is for outside projects.
  something like "Internal Approval" and "Stakeholder Approval"
- [ ] <a href='#crh-comment-Pull 005bfa7ac3b6ec97e0db347241dfbd348300f94f business_requirement/models/business.py 226'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#discussion_r66717616'>File: business_requirement/models/business.py:L1-437</a></b>
- <a href='https://github.com/elicoidal'></a> in v8 this is pending rewriting/refactoring (currently not good to rewrite the whole function in this module)
- [ ] <a href='#crh-comment-Pull 005bfa7ac3b6ec97e0db347241dfbd348300f94f business_requirement/models/business.py 20'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#discussion_r66717660'>File: business_requirement/models/business.py:L1-437</a></b>
- <a href='https://github.com/elicoidal'></a> normally this case is impossible (we have have the discussion in v8)
- [ ] <a href='#crh-comment-Pull 005bfa7ac3b6ec97e0db347241dfbd348300f94f business_requirement/wizards/mail_compose_message.py 8'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#discussion_r66717785'>File: business_requirement/wizards/mail_compose_message.py:L1-88</a></b>
- <a href='https://github.com/elicoidal'></a> same issue here: we should not rewrite this function
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/10#issuecomment-225416384'>General Comment</a></b>
- <a href='https://github.com/elicoidal'></a> @jbeficent Thanks for this PR. I think we might want to finish all rework in v8 before we merge this one because many topics might apply in both cases
- <a href='https://github.com/dreispt'></a> I agree. This PR should be rebased on the 8.0 final version, once its merged.
- <a href='https://github.com/elicoidal'></a> @jbeficent could you review the work done in current branch 8.0 for business_requirement merged module and adapt this PR accordingly?
- <a href='https://github.com/elicoidal'></a> @jbeficent I close this PR as the structure has greatly evolved in v8 since this initial PR.
  We should start a migration process at this point of the v8 repo.

<a href='https://www.codereviewhub.com/OCA/business-requirement/pull/10?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/OCA/business-requirement/pull/10?approve=1'></a>&nbsp;<a href='https://github.com/OCA/business-requirement/pull/10'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: business_requirement/README.rst [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+   :alt: License: AGPL-3
+
+====================
+Business Requirement
+====================
+
+Introduction
+^^^^^^^^^^^^
+
+This module is part of a set ("Business Requirement") and provides the basic
```
Comment (elicoidal) [inline business_requirement/README.rst:12]: I have pending for v8 a remodeling of the intro of the README. Later on we might adapt it to v9 as well

File: business_requirement/README.rst [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+* Integration with earned-value module.
+* Improve the followers details (eg: depending on the stages)
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/
+project/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback `here <https://github.com/OCA/
```
Comment (elicoidal) [inline business_requirement/README.rst:250]: simplify following latest README template

File: business_requirement/README.rst [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+
+* Eric Caudal <eric.caudal@elico-corp.com>
+* Alex Duan <alex.duan@elico-corp.com>
+* Xie XiaoPeng <xie.xiaopeng@elico-corp.com>
+* Victor M. Martin <victor.martin@elico-corp.com>
+
```
Comment (elicoidal) [inline business_requirement/README.rst:271]: add your name

File: business_requirement/models/business.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    def _sub_br_count(self):
+        self.sub_br_count = len(self.business_requirement_ids)
+
+    @api.model
+    def _get_states(self):
+        states = [
+            ('draft', 'Draft'),
+            ('confirmed', 'Confirmed'),
+            ('approved', 'Approved'),
+            ('customer_approval', 'Customer Approval'),
```
Comment (elicoidal) [inline business_requirement/models/business.py:179]: I wonder whether we should be more generic as "customer approval" is for outside projects.
something like "Internal Approval" and "Stakeholder Approval"

File: business_requirement/models/business.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+    @api.one
+    def action_button_cancel(self):
+        self.write({'state': 'cancel'})
+
+    @api.one
+    def action_button_drop(self):
+        self.write({'state': 'drop'})
+
+    @api.cr_uid_ids_context
+    def message_post(self, cr, uid, thread_id, body='', subject=None,
```
Comment (elicoidal) [inline business_requirement/models/business.py:226]: in v8 this is pending rewriting/refactoring (currently not good to rewrite the whole function in this module)

File: business_requirement/models/business.py [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+class BusinessRequirement(models.Model):
+    _inherit = ['mail.thread', 'ir.needaction_mixin']
+    _name = "business.requirement"
+    _description = "Business Requirement"
+    _order = 'id desc'
+
+    @api.model
+    def _get_default_company(self):
+        company_id = self.env.user._get_company()
+        if not company_id:
```
Comment (elicoidal) [inline business_requirement/models/business.py:20]: normally this case is impossible (we have have the discussion in v8)

File: business_requirement/wizards/mail_compose_message.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+# © 2016 Elico Corp (https://www.elico-corp.com).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models
+
+
+class MailComposeMessage(models.TransientModel):
```
Comment (elicoidal) [inline business_requirement/wizards/mail_compose_message.py:8]: same issue here: we should not rewrite this function

**Comment by elicoidal** [comment]
@jbeficent Thanks for this PR. I think we might want to finish all rework in v8 before we merge this one because many topics might apply in both cases

**Comment by dreispt** [comment]
I agree. This PR should be rebased on the 8.0 final version, once its merged.

**Comment by elicoidal** [comment]
@jbeficent could you review the work done in current branch 8.0 for business_requirement merged module and adapt this PR accordingly?

**Comment by elicoidal** [comment]
@jbeficent I close this PR as the structure has greatly evolved in v8 since this initial PR.
We should start a migration process at this point of the v8 repo.

**Pull request closed without merging**